Polynomial chaos cross-validation (#2330)
openturns/openturns: python/doc/theory/meta_modeling/cross_validation.rst

is the domain of the input parameters :math:`\vect{x}`.
Let :math:`\metamodel` be a surrogate model of :math:`\physicalmodel`, i.e.
an approximation of the function.
Once a metamodel :math:`\metamodel`
of the original numerical model :math:`\physicalmodel` has been
built up, we may estimate the mean squared error, i.e. the
discrepancy between the response surface and the true model response
in terms of the weighted :math:`L_2`-norm:

.. math::
   \operatorname{MSE} \left(\metamodel\right)
   & = \left\| \physicalmodel(\vect{x}) - \metamodel(\vect{x}) \right\|_{L_2}^2(\set{X}) \\
   & = \int_{\set{X}} \left[ \physicalmodel(\vect{x}) - \metamodel(\vect{x}) \right]^2  \pdf d\vect{x}

where :math:`\probabilitydensity` is the probability density function
of the random vector :math:`\vect{X}`.
In this section, we present the cross-validation of linear least squares
model, as presented in [wang2012]_ page 485.

The fraction of variance unexplained is:

.. math::

    \operatorname{FVU}
    = \frac{\operatorname{MSE}\left(\metamodel\right)}{\Var{Y}}

where :math:`\Var{Y}` is the variance of the random output :math:`Y = \physicalmodel(\vect{X})`.

> The fraction of variance unexplained is a relative mean squared error.

Author: No : https://en.wikipedia.org/wiki/Fraction_of_variance_unexplained

.. figure:: ../../_static/train_test_cross_validation.png
    :align: center
    :alt: Validation of a surrogate
    :width: 50%

    *Figure 1.* Validation using one single split.

Let :math:`n \in \Nset` be an integer representing the sample size.
Let :math:`\set{D} \subseteq \set{X}` be a set of :math:`n` independent
observations of the random vector :math:`\vect{X}`:

.. math::
    \set{D} = \left\{\vect{x}^{(1)}, ..., \vect{x}^{(n)} \in \Rset^{n_x}\right\}

and consider the corresponding outputs of the model:

.. math::
    \left\{y^{(1)}, ..., y^{(n)}  \in \Rset\right\}

where:

.. math::
    y^{(j)} = \physicalmodel\left(\vect{x}^{(j)}\right)

for :math:`j = 1, ..., n`.
The Monte-Carlo estimator of the mean squared error is:

.. math::
    \widehat{\operatorname{MSE}}\left(\metamodel\right)
    = \frac{1}{n} \sum_{j = 1}^n \left( y^{(j)} - \metamodel \left(\vect{x}^{(j)}\right) \right)^2

The previous equation can be equivalently expressed depending on
the model since :math:`y^{(j)} = \physicalmodel\left(\vect{x}^{(j)}\right)`.
It seems, however, more consistent to use :math:`y^{(j)}` because the
true model :math:`g` is unknown (otherwise we would not use a

> surrogate).

Author: Ok. So let's keep this as is.

may underestimate the true value of the mean squared error.
In order to create a test set independent from the training set, a
simple method is to split the data set into two parts.
The drawback of this method is that this reduces the size of the training
set, so that the mean squared error evaluated on the test set can be pessimistic.
The leave-one-out (LOO) and K-Fold cross validation methods presented in the next sections
use all the data available.

Naive and fast cross-validation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

As seen in the previous section, the simplest method splits the data
into a training set and a test set.
Moreover, provided these two sets are independent, then the estimate
of the error is unbiased.
In order to use all the available data instead of a subset of it,
two other estimators have been set up:
the leave-one-out and K-Fold estimators, which are the topic of the next
sections.

When implemented naively, these methods may require to build many surrogate models,
which can be time-consuming.
Fortunately, there are *shortcuts* for many surrogate models
including linear least squares and splines (and others).

> For a linear least squares model, some methods uses the

Reviewer: ```suggestion
For a linear least squares model, some methods use the
```

Sherman-Morrisson-Woodbury formula to get updates of the inverse Gram matrix,
as we are going to see later in this document.

> This makes it possible to evaluate easily the meta model errors

Reviewer: ```suggestion
This makes it possible to easily evaluate metamodel errors
```

of a linear least squares model.

Leave-one-out cross-validation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In this section, we present the naive leave-one-out error estimator,
also known as jackknife in statistics.
Let :math:`\metamodel^{(-j)}` be the surrogate model estimated from the
leave-one-out experimental design :math:`\set{D} \setminus \{\vect{x}^{(j)}\}`.
This is the experimental design where the :math:`j`-th observation
:math:`\vect{x}^{(j)}` is set aside.
The corresponding set of observation indices is:

.. math::
    \set{S}_{-j} = \{1, ..., j - 1, j + 1, ..., n\},

the corresponding input observations are:

.. math::
    \set{D} \setminus \{\vect{x}^{(j)}\}
    = \left\{\vect{x}^{(1)}, ..., \vect{x}^{(j - 1)},
    \vect{x}^{(j + 1)}, ..., \vect{x}^{(n)}\right\}

>     = \left\{\vect{x}^{(j)}, \; j \in \set{S}_{-j}\right\}.

Reviewer: ```suggestion
    = \left\{\vect{x}^{(j)}, \; j \in \set{S}_{-j}\right\}
```
The sentence is not finished.

:math:`\vect{x}^{(j)}` and its leave-one-out prediction (see [blatman2009]_
eq. 4.26 page 85):

.. math::

    \Delta^{(j)}
    = y^{(j)} - \metamodel^{(-j)}\left(\vect{x}^{(j)}\right)

We repeat this process for all observations in the experimental
design and obtain the predicted residuals
:math:`\Delta^{(j)}` for :math:`j = 1, \dots, n`.
Finally, the LOO mean squared error estimator is:

.. math::

    \widehat{\operatorname{MSE}}_{LOO}
    & =  \frac{1}{n} \sum_{j = 1}^n \left( \Delta^{(j)} \right)^2

One of the drawbacks of the naive method is that it may require
to estimate :math:`n` different surrogate models.
If :math:`n` is large or if training each surrogate model is costly,
then the leave-one-out method can be impractical.
If, however, the surrogate model is based on the linear least squares method,
then the leave-one-out error may be computed much more efficiently, as
shown in the next section.

Fast leave-one-out cross-validation of a linear model
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In this section, we present the fast leave-one-out error estimator
of a linear least squares model.

> In the special case of of a linear least squares model, [stone1974]_ (eq. 3.13 page 121)

Author: ```suggestion
In the special case of a linear least squares model, [stone1974]_ (eq. 3.13 page 121)
```

showed that the leave-one-out residuals have an expression which depends on the diagonal
of the projection matrix.
In this case, the evaluation of the leave-one-out mean squared error involves the
multiplication of the raw residuals by a correction which involves the leverages
of the model.

> This method makes it possible to evaluation directly the mean squared error without

Reviewer: ```suggestion
This method makes it possible to directly evaluate the mean squared error without
```

for any :math:`\vect{x} \in \set{X}` where :math:`\vect{a} \in \Rset^{n_x + 1}`
is the vector of parameters.
Let :math:`\vect{y} \in \Rset^n` be the vector of output observations:

.. math::
    \vect{y} = \left(y^{(1)}, ..., y^{(n)} \right)^T.

The goal of the least squares method is to estimate the coefficients
:math:`\vect{a}` using the vector of observations :math:`\vect{y}`.
The output vector from the linear model is:

.. math::

    \vect{y} =  \boldsymbol{D} \vect{a}

for any :math:`\vect{x} \in \set{X}` where
:math:`\boldsymbol{D} \in \Rset^{n \times (n_x + 1)}` is the
design matrix.
For a linear model, the columns of the design matrix correspond
to the input parameters and the rows correspond to the observations:

.. math::
    \boldsymbol{D}
    =
    \begin{pmatrix}
    1      & x_1^{(1)} & \ldots  & x_{n_x}^{(1)} \\
    \vdots &           &         & \vdots \\
    1      & x_1^{(n)} & \ldots  & x_{n_x}^{(n)}
    \end{pmatrix}.

In the previous equation, notice that the design matrix depends on the
experimental design :math:`\set{D}`.

Assume that the matrix :math:`\boldsymbol{D}` has full rank.
The solution of the linear least squares problem is

> given by the normal equations ([Bjorck1996]_ eq. 1.1.15 page 6):

Author: ```suggestion
given by the normal equations (see [Bjorck1996]_ eq. 1.1.15 page 6):
```

.. math::

    \widehat{\vect{a}} = \left(\boldsymbol{D}^T \boldsymbol{D} \right)^{-1} \boldsymbol{D}^T \vect{y}.

The linear surrogate model is the linear model with estimated coefficients:

.. math::

    \metamodel(\vect{x}) = \hat{a}_0 + \sum_{i=1}^{n_x} \hat{a}_i x_i.

The vector of predictions from the surrogate model is:

.. math::

    \widehat{\vect{y}} = \boldsymbol{D} \widehat{\vect{a}}

for any :math:`\vect{x} \in \set{X}` where :math:`\widehat{\vect{a}}` is the
estimate from linear least squares.
We substitute the estimator in the previous equation and
get the value of the surrogate linear model:

.. math::

    \widehat{\vect{y}}
    = \boldsymbol{D} \left(\boldsymbol{D}^T \boldsymbol{D} \right)^{-1} \boldsymbol{D}^T \vect{y}

> Let :math:`\boldsymbol{H} \in \Rset^{n \times n}` be the projection ("hat") matrix ([wang2012]_ eq. 16.8 page 472):

Author: ```suggestion
Let :math:`\boldsymbol{H} \in \Rset^{n \times n}` be the projection ("hat") matrix (see [wang2012]_ eq. 16.8 page 472):
```

.. math::
    \sum_{j = 1}^{n} h_{jj}  = \operatorname{Tr}(H) = n_x + 1

where :math:`\operatorname{Tr}(H)` is the trace of the hat matrix.
The leverage describes how far away the individual data point is from the centroid
of all data points (see [sen1997]_ page 155).
The equation :eq:`predictionCorrection` implies that if :math:`h_{jj}` is
large (i.e. close to 1), then removing the :math:`j`-th observation
from the training sample changes the residual of the leave-one-out
surrogate model significantly.

Using the equation :eq:`predictionCorrection` avoids to actually build the LOO surrogate.
We substitute the previous expression in the definition of the leave-one-out
mean squared error estimator and get the fast leave-one-out cross validation
error ([wang2012]_ eq. 16.25 page 487):

.. math::
    \widehat{\operatorname{MSE}}_{LOO}
    = \frac{1}{n} \sum_ {j = 1}^n \left( \frac{y^{(j)} -
          \widehat{g}\left(\vect{x}^{(j)}\right)}{1 - h_{jj}} \right)^2

Corrected leave-one-out
~~~~~~~~~~~~~~~~~~~~~~~

A penalized variant of the leave-one-out mean squared error may be used in order to
increase its robustness with respect to overfitting.

> This is done using a criteria which takes into account for the
> number of coefficients compared to the size of the

Reviewer: ```suggestion
This is done using a criterion which takes into account the
number of coefficients compared to the size of the
```

experimental design.
The corrected leave-one-out error is ([chapelle2002]_, [blatman2009]_ eq. 4.38 page 86):

.. math::
    \widehat{\operatorname{MSE}}_{LOO}^{*}
    = \widehat{\operatorname{MSE}}_{LOO} T(n_x, n)

where the penalty factor is:

.. math::
    T(n_x, n)
    = \frac{n}{n - (n_x + 1)}  \left(1 + \frac{\operatorname{Tr} \left( \boldsymbol{C}_{emp}^{-1}  \right) }{n} \right)

where :math:`\boldsymbol{C}_{emp}` is the matrix:

.. math::

    \boldsymbol{C}_{emp} = \frac{1}{n}\boldsymbol{D}^{\textsf{T}}\boldsymbol{D}

and :math:`\operatorname{Tr}` is the trace operator.

K-fold cross-validation
~~~~~~~~~~~~~~~~~~~~~~~

In this section, we present the naive K-Fold cross-validation.
Let :math:`k \in \Nset` be a parameter representing the number of
splits in the data set.
The :math:`k`-fold cross-validation technique relies on splitting the

> data set :math:`\set{D}` into :math:`k` mutually disjoint sub-samples

Author: ```suggestion
data set :math:`\set{D}` into :math:`k` sub-samples
```

:math:`\set{D}_1, \dots, \set{D}_k`, called the *folds*.
The corresponding set of indices:

.. math::

    \set{S}_1 \; \dot{\cup} \; \cdots \; \dot{\cup} \; \set{S}_k
    = \{1, ..., n\}

and the corresponding set of input observations is:

.. math::

>     \set{D} = \set{D}_1 \; \dot{\cup} \; \cdots \; \dot{\cup} \; \set{D}_k.

Author: ```suggestion
    \set{D} = \set{D}_1 \; \cup \; \cdots \; \cup \; \set{D}_k.
```

The next figure presents this type of cross validation.

.. figure:: ../../_static/kfold_validation.png
    :align: center
    :alt: K-Fold validation
    :width: 75%

    *Figure 2.* K-Fold cross-validation.

> The :math:`k` folds are generally chosen so that they are of

Reviewer: ```suggestion
The :math:`k` folds are generally chosen so that they are of
```
```suggestion
The :math:`k` folds are generally chosen to be of
```

approximately equal sizes.
If the sample size :math:`n` is a multiple of :math:`k`, then the
folds can have exactly the same size.

For any :math:`\ell \in \{1, ..., k\}`, let :math:`\metamodel^{(-\set{D}_{\ell})}`
be the surrogate model estimated on the K-fold sample
:math:`\set{D} \setminus \set{D}_{\ell}`.
Let :math:`\Delta^{(\ell, j)}` be defined as the K-Fold residual:

.. math::

    \Delta^{(\ell, j)}
    = y^{(j)} - \widehat{g}^{(-\set{D}_{\ell})} \left(\vect{x}^{(j)}\right)

for :math:`\ell = 1, ..., k` and :math:`j \in \set{S}_{\ell}`.
In the previous equation, the *predicted residual*
:math:`y^{(j)} - \metamodel^{(-\set{D}_{\ell})} \left(\vect{x}^{(j)}\right)`
is the difference between the
evaluation of :math:`\physicalmodel` and the value of the K-Fold surrogate
:math:`\metamodel^{(-\set{D}_{\ell})}` at the point :math:`\vect{x}^{(j)}`.
The local approximation error is estimated on the sample :math:`\set{D}_{\ell}`:

.. math::
   \widehat{\operatorname{MSE}}^{(\ell)}

>    = \frac{1}{n_\ell}  \sum_{j \in \set{S}_\ell} \left( \Delta^{(\ell, j)} \right)^2.

Reviewer: ```suggestion
   = \frac{1}{n_\ell}  \sum_{j \in \set{S}_\ell} \left( \Delta^{(\ell, j)} \right)^2
```

corrected K-Fold residuals:

.. math::

    (\boldsymbol{I}_{\ell} - \boldsymbol{H}_{\ell}) \widehat{\boldsymbol{r}}_{\ell}
    = \boldsymbol{y}_{\ell} - \widehat{\boldsymbol{y}}_{\ell}

where :math:`\boldsymbol{I}_{n_{\ell}} \in \Rset^{n_{\ell} \times n_{\ell}}` is the identity matrix,
:math:`\boldsymbol{y}_{\ell} \in \Rset^{n_{\ell}}` is the vector of output observations in the
:math:`\ell`-th fold:

.. math::

    \boldsymbol{y}_{\ell}
    = \left(y^{(j)}\right)^T_{j \in \set{S}_{\ell}}

and :math:`\widehat{\boldsymbol{y}}_{\ell} \in \Rset^{n_{\ell}}` is the corresponding
vector of output predictions from the linear least squares surrogate model:

.. math::

    \widehat{\boldsymbol{y}}_{\ell}
    = \left(g\left(\vect{x}^{(j)}\right)\right)^T_{j \in \set{S}_{\ell}}.

Then the mean squared error of the :math:`\ell`-th fold is:

.. math::

    \widehat{\operatorname{MSE}}^{(\ell)}
    = \frac{1}{n_{\ell}} \sum_{j \in \set{S}_{\ell}}
    \left(\widehat{\boldsymbol{r}}_{\ell}\right)_j^2.

Then the K-Fold mean squared error is evaluated from equation :eq:`kfoldMean`.

Cross-validation and model selection
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> If a model selection method is used (such as :class:`LARS`), then the fast cross-validation (CV)

Reviewer: ```suggestion
If a model selection method is used (such as :class:`~openturns.LARS`), then the fast cross-validation (CV)
```

method can produce an optimistic estimated error i.e. the true error can
be greater that the estimated error (see [hastie2009]_ section 7.10.2 page 245).

> This is because the fast CV does not take into account for the model selection.

Reviewer: ```suggestion
This is because the fast CV does not take model selection into account.
```

The reason for this behavior is that the model selection produces a set

> of predictors which fits particularly well to the data.

Reviewer: ```suggestion
of predictors which fits the data particularly well.
```

- :class:`~openturns.KFoldSplitter`: uses the K-Fold method
  to split the data set.

> Since the :class:`~openturns.LinearModelResult` is based on linear least
> squares, fast methods are implemented in the :class:`~openturns.experimental.LinearModelValidation`.

Reviewer: ```suggestion
Since :class:`~openturns.LinearModelResult` is based on linear least
squares, fast methods are implemented in :class:`~openturns.experimental.LinearModelValidation`.
```

See :ref:`pce_cross_validation` and :class:`~openturns.FunctionalChaosValidation`
for specific methods for the the cross-validation of a polynomial chaos expansion.

.. topic:: API:

    - See :class:`~openturns.MetaModelValidation`

>     - See :class:`~openturns.LinearModelValidation`

Reviewer 2: ```suggestion
    - See :class:`~openturns..experimental.LinearModelValidation`
```

Author: This is fixed.

>     - See :class:`~openturns.FunctionalChaosValidation`

Reviewer 2: ```suggestion
    - See :class:`~openturns..experimental.FunctionalChaosValidation`
```

Author: This is fixed.
